### PR TITLE
feat: eip1559 fees

### PR DESCRIPTION
**Pull request by kaladinlight**
## Description

Update to use EIP-1559 fee where possible in swapper

- approvals (zrx, thorswap, cowswap, lifi, univ2, fox-farming)
- sends/swaps (zrx, thorswap, cowswap, univ2, fox-farming)

* using `average` eip1559 gas price and `fast` legacy gas price

Lifi will be a follow up for the rest of the flow outside of approval.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Risk

high - this touches a lot of the gas logic which is very fragmented across the code base with possibility of a missed edge case

## Testing

- zrx full functionality
- thorswap full functionality
- cowswap full functionality
- lifi full functionality
- uniV2 full functionality
- fox-farming full functionality

### Engineering

:point_up: 

### Operations

:point_up: 

## Screenshots (if applicable)

**ZRX**:
- approval: 
image
https://etherscan.io/tx/0x3f20cf391a8c00b6f12f7e00640c1ea6ed63ad7b118d38f9a606e646d6e7f798
- swap:
image
https://etherscan.io/tx/0x3f1e142b9128d16d0728614a96fd6c41998caa41a263ab026ce9ed4f0db2ab8d

**Thor**:
- approval:
image
https://etherscan.io/tx/0x62f387ee148475665d8b74affe7bbd6aabffb96d1fccdeb335d5eb10d8148b68
- swap:
image
https://etherscan.io/tx/0x23a738adf80157026995bd8506431ab83337caf6951151deed2a50679343dbfe

**FoxFarming**:
- approval:
image
https://etherscan.io/tx/0xbacb24242a6fbbcb1b2e3fd316270c1b763afed48ac21bb0ad1eb4b0f8d4f662
- stake:
image
https://etherscan.io/tx//0x3bbbf5dbc21410978ae23d09f75c640fcd76f204f84b7c4849d8942b08abd87b
- claim:
image
https://etherscan.io/tx/0x450b251f874a4938acbe7f4bf1aab8b967370161d8ae6858e6ebc6944a52d77f
- unstake
image
https://etherscan.io/tx//0x7e8e43ae424dd07367fc32963550bacf636ef78738ad95c903b89698c463d7bf

**UniV2**:
- approval:
image
https://etherscan.io/tx/0x8b243f7de0f94100349c5a3f5c502d217fbfcf256224f09a7d637604aafd05cd
- deposit:
image
https://etherscan.io/tx/0xa9a0412ee6c1080361de8fa5ea4f9f8e691a0dfc6211df2f88d43a7cbf4f51de
- withdraw:
image
https://etherscan.io/tx/0x6ce7c41042b80f48b6711bca8729da543ff290b268636c03096fe5cedfb97a26

### CODE REVIEW

File: package.json
Comment (kaladinlight) [inline package.json:276]: turns out react-app-rewired doesn't use `jest.config.js`

File: packages/chain-adapters/src/evm/EvmBaseAdapter.ts
Comment (gomesalexandre) [inline packages/chain-adapters/src/evm/EvmBaseAdapter.ts:550]: ty, that was my bad copying this one from web (which itself could do with the same love using `numberToHex`)

File: src/features/defi/helpers/utils.ts
Comment (gomesalexandre) [inline src/features/defi/helpers/utils.ts:50]: Noice, that will help narrowing `ETHWallet` down without having to run this one over and over everytime we get fees I reckon

File: src/lib/swapper/swappers/CowSwapper/cowApprove/cowApprove.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/CowSwapper/cowApprove/cowApprove.ts:16]: Obviously not for this PR but would be nice to make a discriminated union of `approvalAmount` / `infinite: boolean` (which doesn't exist yet)

File: src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts:146]: Feels like too much magic with this destructuring but why not

File: src/lib/swapper/swappers/ThorchainSwapper/utils/txFeeHelpers/evmTxFees/getEvmTxFees.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/txFeeHelpers/evmTxFees/getEvmTxFees.ts:22]: Inconsistent destructuring of average vs. reassignment to `fee`

File: src/lib/swapper/swappers/ZrxSwapper/ZrxSwapper.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ZrxSwapper/ZrxSwapper.ts:90]: noice

**Review by gomesalexandre — COMMENTED** [review]
Untested but conceptually looks excellent so far

File: src/features/defi/helpers/utils.ts
Comment (kaladinlight) [reply src/features/defi/helpers/utils.ts:50]: Yeah pretty much, cleans up usage and does what it needs to do

File: src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts
Comment (kaladinlight) [reply src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts:146]: meh will pull it out just for clarity

File: src/lib/swapper/swappers/CowSwapper/cowApprove/cowApprove.ts
Comment (kaladinlight) [reply src/lib/swapper/swappers/CowSwapper/cowApprove/cowApprove.ts]: Simplified it a bit more, not sure if discrimination is worth

File: src/features/defi/helpers/utils.ts
Comment (gomesalexandre) [inline src/features/defi/helpers/utils.ts:51]: "display purposes" might be a bit of a stretch, this is really used as view-layer logic, hence will block users from continuing if they don't have enough funds for the fast fees - which isn't a bad heuristic safety-wise, always good to plan a buffer.

I'd probably revamp that comment to say "view-layer logic purposes" or similar though. At first reading this, I thought of this as "display the fast fee, but use the average fee under the hood for everything else including has enough native token for gas calculations", but double-checking consumption proved otherwise

File: src/features/defi/providers/univ2/components/UniV2Manager/Deposit/components/Deposit.tsx
Comment (gomesalexandre) [inline src/features/defi/providers/univ2/components/UniV2Manager/Deposit/components/Deposit.tsx]: Can get `txFee` on the right side of the short circuit since this is the only property we use from `FeeData` here

File: src/features/defi/providers/univ2/hooks/useUniV2LiquidityPool.ts
Comment (gomesalexandre) [inline src/features/defi/providers/univ2/hooks/useUniV2LiquidityPool.ts:474]: Can abstract to a var so we only call this once if you feel likt it

**Review by gomesalexandre — COMMENTED** [review]
Tested with a lil help from Tenderly:

### UNI-V2 approvals

#### FOX

image

#### UNI (More logic in their `approve()` method which makes us require a bit of buffer currently, see https://github.com/Uniswap/governance/blob/eabd8c71ad01f61fb54ed6945162021ee419998e/contracts/Uni.sol#L151-L155)

image

### FOX Farming

https://etherscan.io/tx/0xbb33a4111850fda35d3811fc9d9d483081e47332b3eca09494de34d0561d8fcb

### ZRX Swapper using the 100000 default `APPROVAL_GAS_LIMIT`

image

Not for this PR, but we may want to not hardcode a gas limit for approvals anymore in the future, now that we have more sane estimations

### CoW swapper using the default `APPROVAL_GAS_LIMIT`

image

### THOR Swapper

image

## Sends/swaps

### THOR Swapper using the 100000 `THOR_EVM_GAS_LIMIT`

image

### CoW swapper

Unable to test with Tenderly since this uses signatures and would imply broadcasting an actual Tx - this diff should produce no changes here

### ZRX

Unable to test without testing funds, though running through the gas limit / usage of previous similar calls over the last minutes for the `sellToUniswap()` method, our `gasLimit` has enough buffer here ✅ will obviously need to be tested by ops

Didn't notice any regressions here, according to Tenderly which gives us the same actual deterministic guarantees as broadcasting Txs.
Will obviously need some ops love of legacy/EIP-1559 after this goes in, and ideally https://github.com/shapeshift/hdwallet/pull/614 to go in and be bumped here as well to ensure MM uses EIP-1559 fees, but conceptually looks good to me functionally so far!

File: src/features/defi/helpers/utils.ts
Comment (kaladinlight) [reply src/features/defi/helpers/utils.ts]: https://github.com/shapeshift/web/pull/4324/commits/967fcf283682f83ae35dc3112cdbf363df1cbc31

File: src/features/defi/providers/univ2/hooks/useUniV2LiquidityPool.ts
Comment (kaladinlight) [reply src/features/defi/providers/univ2/hooks/useUniV2LiquidityPool.ts:474]: not sure I follow

Comment (gomesalexandre) [reply src/features/defi/providers/univ2/hooks/useUniV2LiquidityPool.ts:474]: Disregard me, GitHub collapsing rugged me cognitively

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4324** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4324" target="_blank">Graphite</a>  👈
    * **PR #3508** <a href="https://app.graphite.dev/github/pr/shapeshift/web/3508" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4324?utm_source=stack-comment).

**Review by gomesalexandre — APPROVED** [review]
LGTM after the latest changes and the UNI-V2 `addLiquidityEth` ETH value fix - re-tested in a poile

**Pull request merged**
